### PR TITLE
Add simple integration testing on vm invoke tasks

### PR DESCRIPTION
**Pull request by KevinFairise2**
What does this PR do?
---------------------

Add simple integration tests on VM invoke tasks

Which scenarios this will impact?
-------------------

Motivation
----------

Be sure we do not break our invoke tasks when making changes

Additional Notes
----------------

### CODE REVIEW

File: tests/invoke-tasks-test.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+aws ec2 delete-key-pair --key-name ci.test-infra-definitions.test-key-$CI_PIPELINE_ID
+aws ec2 import-key-pair --key-name ci.test-infra-definitions.test-key-$CI_PIPELINE_ID --public-key-material fileb://key-pair.pub
+
+echo "Running inv setup"
+printf "$ENV\nci.test-infra-definitions.test-key-$CI_PIPELINE_ID\nN\nkey-pair.pub\ntest-ci\n00000000000000000000000000000000\n0000000000000000000000000000000000000000\n" | inv setup
+setup_exit_code=$?
+
+echo "Running inv create-vm"
+export PULUMI_CONFIG_PASSPHRASE=$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c10)
+inv create-vm -s ci-integration-testing-$CI_PIPELINE_ID --private-key-path $KEYPAIR_PATH
```
Comment (ogaca-dd) [inline tests/invoke-tasks-test.sh:20]: Is there a reason to write the test in bash and not in Go? Writing in Go provides more control especially if we want to write more tests later.

File: tasks/setup.py [hunk 2]
```diff
@@ -8,9 +8,10 @@
 from invoke.tasks import task
 
 from .config import Config, get_full_profile_path, get_local_config
-from .tool import ask, info, is_windows, warn
+from .tool import ask, info, is_windows, warn, is_ci
 
-available_aws_accounts = ["agent-sandbox", "sandbox"]
+
+available_aws_accounts = ["agent-sandbox", "sandbox"] if not is_ci() else ["agent-qa"]
```
Comment (ogaca-dd) [inline tasks/setup.py:14]: I would just add `"agent-qa"` to the list, ie, `available_aws_accounts = ["agent-sandbox", "sandbox", "agent-qa"]` as a user may want to run the setup for `"agent-qa"`.

File: tasks/setup.py [hunk 3]
```diff
@@ -38,11 +39,12 @@ def setup(_: Context) -> None:
     setupAgentConfig(config)
 
     config.save_to_local_config()
-    cat_profile_command = f"cat {get_full_profile_path()}"
-    pyperclip.copy(cat_profile_command)
-    print(
-        f"\nYou can run the following command to print your configuration: `{cat_profile_command}`. This command was copied to the clipboard\n"
-    )
+    if not is_ci():
```
Comment (ogaca-dd) [inline tasks/setup.py:42]: What happens if you remove this condition?

File: tasks/tool.py [hunk 4]
```diff
@@ -85,7 +89,9 @@ def get_stack_json_outputs(ctx: Context, full_stack_name: str) -> Any:
 
 
 def get_aws_wrapper(aws_account: str) -> str:
-    return f"aws-vault exec sso-{aws_account}-account-admin"
+    if tool.is_ci():
```
Comment (ogaca-dd) [inline tasks/tool.py:92]: I am wondering if we should add instead a command line flag named `with-aws-vault`? What do you think?

File: tasks/vm.py [hunk 5]
```diff
@@ -51,6 +52,9 @@ def create_vm(
     if ami_id is not None:
         extra_flags["ddinfra:osAmiId"] = ami_id
 
+    if private_key_path is not None:
```
Comment (ogaca-dd) [inline tasks/vm.py:55]: Can you add instead the ssh key to the ssh agent so the code is the same in both cases?

File: tests/invoke-tasks-test.sh [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+KEYPAIR_PATH=key-pair
+ENV=agent-qa
+
+echo "Creating ssh key pair"
+
+ssh-keygen -t ed25519 -C "Temporary key pair used for test-infra-definitions integration tests" -f key-pair
+
+echo "Import key pair on AWS"
+
+aws ec2 delete-key-pair --key-name ci.test-infra-definitions.test-key-$CI_PIPELINE_ID
```
Comment (ogaca-dd) [inline tests/invoke-tasks-test.sh:11]: How is it done in the `datadog-agent` repository when running e2e tests?

File: tests/invoke-tasks-test.sh [hunk 7]
```diff
@@ -25,7 +25,7 @@ else
     failed=true
 fi
 
-if [[ $destroy_exit_code -eq 0 ]]; then
+if [[ $create_exit_code -eq 0 ]]; then
```
Comment (ogaca-dd) [inline tests/invoke-tasks-test.sh:28]: Is this file still needed?

File: tests/invoke_test.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestInvokeVM(t *testing.T) {
+
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	setupCmd := exec.Command("invoke", "setup")
+	setupCmd.WaitDelay = 5 * time.Second
```
Comment (ogaca-dd) [inline tests/invoke_test.go:20]: Can you explain the purpose of this line?

Comment (KevinFairise2) [reply tests/invoke_test.go:20]: Some tests I did, but it is useless now. Removing it

File: tests/invoke-tasks-test.sh (hunk 7 above)
Comment (KevinFairise2) [reply tests/invoke-tasks-test.sh]: No, deleted

File: tests/invoke-tasks-test.sh (hunk 6 above)
Comment (KevinFairise2) [reply tests/invoke-tasks-test.sh]: Just took what was done in `datadog-agent` to have things done the same way

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -18,4 +18,4 @@ jobs:
         run: go build ./...
       
       - name: Test
-        run: go test -v ./...
+        run: go test -v $(go list ./... | grep -v /integration-tests)
```
[the remark this answers is not in the document]
Comment (KevinFairise2) [reply .github/workflows/build.yml:21]: If we try to run it here. It does not work since the integration tests must be ran in an image with more tooling (Pulumi, invoke, ...)

File: integration-tests/invoke_test.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+	_, err = io.WriteString(stdin, "0000000000000000000000000000000000000000\n")
+	require.NoError(t, err, "Failed to write user input to stdin")
+	err = stdin.Close()
+	require.NoError(t, err, "Failed to close stdin pipe")
+
+	err = setupCmd.Wait()
+	require.NoError(t, err, "Error found: %s %s", setupStdout.String(), setupStderr.String())
+	require.Contains(t, setupStdout.String(), "Configuration file saved at", "If setup succeeded, last message should contain 'Configuration file saved at'")
+
+	createCmd := exec.Command("invoke", "create-vm", "--stack-name", fmt.Sprintf("integration-testing-%s", os.Getenv("CI_PIPELINE_ID")), "--no-copy-to-clipboard", "--no-use-aws-vault")
```
[the remark this answers is not in the document]
Comment (KevinFairise2) [reply integration-tests/invoke_test.go:48]: Before running `create-vm`, `setup` must be called. I could create a test to test only `setup` and another one to test `create-vm` but it would still require to run `setup` before.

File: integration-tests/invoke_test.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	createCmd.Stdout = &createStdout
+	createCmd.Stderr = &createStderr
+	err = createCmd.Run()
```
[the remark this answers is not in the document]
Comment (KevinFairise2) [reply integration-tests/invoke_test.go:51]: Good to know I will make the change!

File: requirements.txt [hunk 12]
```diff
@@ -2,5 +2,6 @@ invoke==2.2.0
 termcolor==2.3.0
 pyyaml==6.0.1
 pyperclip==1.8.2
-pydantic==2.1.1 ; python_version >= '3.10'
-pydantic==1.10.9 ; python_version < '3.10'
\ No newline at end of file
+# TODO Fix error raised by pydantic when using version 2.1.1
```
[the remark this answers is not in the document]
Comment (KevinFairise2) [reply requirements.txt:5]: When the test is ran in the CI it uses python 3.10, and with pydantic 2.1.1 our tasks raise errors. I could reproduce the same errors locally. So I decided to use pydantic 1.10.9 even with python >= 3.10. And we could investigate this in another card

File: tasks/deploy.py [hunk 13]
```diff
@@ -123,8 +125,9 @@ def _deploy(
         up_flags += " --debug"
 
     _create_stack(ctx, stack_name, global_flags)
-
-    cmd = f"{tool.get_aws_wrapper(aws_account)} -- pulumi {global_flags} up --yes -s {stack_name} {up_flags}"
+    cmd = f"pulumi {global_flags} up --yes -s {stack_name} {up_flags}"
+    if use_aws_vault:
```
[the remark this answers is not in the document]
Comment (KevinFairise2) [reply tasks/deploy.py:129]: I will set `use_aws_vault` default value to `true` to be sure it is either defined by the user or `true`

File: .github/workflows/build.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply .github/workflows/build.yml:21]: I think it worths adding a comment. What do you think?

[the remark this answers is not in the document]
Comment (KevinFairise2) [reply .github/workflows/build.yml:21]: Sure!

**Review by ogaca-dd — APPROVED** [review]

**Review by ogaca-dd — APPROVED** [review]

File: .gitlab-ci.yml [hunk 14]
```diff
@@ -1,9 +1,36 @@
 variables:
   IMAGE_REPOSITORY: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/test-infra-definitions
+  KUBERNETES_SERVICE_ACCOUNT_OVERWRITE: "test-infra-definitions"
 
 stages:
+  - test
```
Comment (pducolin) [inline .gitlab-ci.yml:6]: **💭 thought**
Why did you choose to test before building the runner image ? Egg and chicken situation here, as you would need to update the runner image tag as soon as you update a version.
